crates/ide_assists/src/handlers/convert_into_to_from.rs

```diff
@@ -0,0 +1,355 @@
+use ide_db::{
+    helpers::{mod_path_to_ast, FamousDefs},
+    traits::resolve_target_trait,
+};
+use syntax::ast::{self, AstNode, NameOwner};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: convert_into_to_from
+//
+// Converts an Into impl to an equivalent From impl.
+//
+// ```
+// # //- /lib.rs crate:core
+// # pub mod convert { pub trait Into<T> { pub fn into(self) -> T; } }
+// # //- /lib.rs crate:main deps:core
+// # use core::convert::Into;
+// impl $0Into<Thing> for usize {
+//     fn into(self) -> Thing {
+//         Thing {
+//             b: self.to_string(),
+//             a: self
+//         }
+//     }
+// }
+// ```
+// ->
+// ```
+// # use core::convert::Into;
+// impl From<usize> for Thing {
+//     fn from(val: usize) -> Self {
+//         Thing {
+//             b: val.to_string(),
+//             a: val
+//         }
+//     }
+// }
+// ```
+pub(crate) fn convert_into_to_from(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let impl_ = ctx.find_node_at_offset::<ast::Impl>()?;
+    let src_type = impl_.self_ty()?;
+    let ast_trait = impl_.trait_()?;
+
+    let module = ctx.sema.scope(impl_.syntax()).module()?;
+
+    let trait_ = resolve_target_trait(&ctx.sema, &impl_)?;
+    if trait_ != FamousDefs(&ctx.sema, Some(module.krate())).core_convert_Into()? {
+        return None;
+    }
+
+    let src_type_path = {
+        let src_type_path = src_type.syntax().descendants().find_map(ast::Path::cast)?;
```

Comment:
I think you should be able to just match on the type here like was done below for `ast_trait`?

Reply:
My first version just had a match for this, but that didn't work for `impl Into<Foo> for &Something` - so updated to this to save having to manually recurse through the various `ast::Type` possibilities.  Can update to do that instead if you'd prefer though?

Reply:
Wouldn't this create incorrect impls for stuff like references then? But I guess this should ideally be done by using the HIR more instead of the AST, though those changes should better be done in a different PR then. So I guess this is fine for now

Reply:
There are probably some cases where this will do the wrong thing, although I'm not sure how common they are.
 
I'm doing this so that I have some type to replace `Self` with if it appears in the body.  The main case I'm trying to cover is `match self { Self::X => {..}}` - which is only valid rust if `Self` is an enum.  If `Self` is a reference type this match statement wouldn't be valid in the first place.  So I wouldn't expect this value to be used in that case - but wanted to make sure I had _something_ to replace `Self` with if I do encounter it - leaving it as `Self` would definitely be wrong, and doesn't look like there's a way to error out of assists once the user has requested them.

I guess one possible failure case for this would be:

```
impl Into<String> for &SomeType {
  fn into(self) -> String {
    let x: Self = self.clone()
    ...
  }
}
```

Definitely going to do the wrong thing with the `x: Self` there.  Not sure if that's a particularly common thing to write, but happy to improve if you've got any suggestions?

Reply:
Ah now I see, ye I think this problem is fine. I don't really see how that can be prevented too easily but I also doubt this shouldn't happen frequently to people.

bors r+